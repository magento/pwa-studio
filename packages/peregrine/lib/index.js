import * as RestApi from './RestApi';
import * as Util from './util';

// hooks
export { useEventListener } from './hooks/useEventListener';
export { useCarousel } from './hooks/useCarousel';
export { useDropdown } from './hooks/useDropdown';
export { usePagination } from './hooks/usePagination';
export { useRestApi } from './hooks/useRestApi';
export { useRestResponse } from './hooks/useRestResponse';
export { useScrollLock } from './hooks/useScrollLock';
export { useSearchParam } from './hooks/useSearchParam';
export { useSort } from './hooks/useSort';
export { useTypePolicies } from './hooks/useTypePolicies';

export {
    WindowSizeContextProvider,
    useWindowSize
} from './hooks/useWindowSize';
export { getToastId, useToasts, ToastContextProvider } from './Toasts';

// store
export { enhancer, reducers } from './store';

// components
export { default as ContainerChild } from './ContainerChild';
export { default as List, Items, Item } from './List';
export { default as Price } from './Price';
export { default as Router } from './Router';
export {
    default as PeregrineContextProvider
} from './PeregrineContextProvider';

// misc
export { RestApi };
export { Util };
export { default as createTestInstance } from './util/createTestInstance';
