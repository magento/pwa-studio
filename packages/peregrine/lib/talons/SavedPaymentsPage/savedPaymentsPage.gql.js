import { gql } from '@apollo/client';

export const SavedPaymentsFragment = gql`
    fragment SavedPaymentsFragment on CustomerPaymentTokens {
        items {
            details
            public_hash
            payment_method_code
        }
    }
`;

export const GET_SAVED_PAYMENTS_QUERY = gql`
    query GetSavedPayments {
        customerPaymentTokens {
            ...SavedPaymentsFragment
        }
    }
    ${SavedPaymentsFragment}
`;

export default {
    getSavedPaymentsQuery: GET_SAVED_PAYMENTS_QUERY
};
