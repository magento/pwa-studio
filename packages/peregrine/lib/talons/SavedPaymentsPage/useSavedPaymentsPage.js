import { useEffect } from 'react';
import { useHistory } from 'react-router-dom';

import { useQuery } from '@apollo/client';

import { useAppContext } from '@magento/peregrine/lib/context/app';
import { useUserContext } from '@magento/peregrine/lib/context/user';

import mergeOperations from '@magento/peregrine/lib/util/shallowMerge';
import defaultOperations from './savedPaymentsPage.gql';

export const normalizeTokens = responseData => {
    const paymentTokens =
        (responseData && responseData.customerPaymentTokens.items) || [];

    return paymentTokens.map(
        ({ details, public_hash, payment_method_code }) => ({
            // details is a stringified object.
            details: JSON.parse(details),
            public_hash,
            payment_method_code
        })
    );
};
/**
 * This talon contains logic for a saved payment page component.
 * It performs effects and returns prop data for rendering the component.
 *
 * @function
 *
 * @param {Object} props
 * @param {SavedPaymentsPageQueries} props.operations GraphQL queries
 *
 * @returns {SavedPaymentsPageTalonProps}
 *
 * @example <caption>Importing into your project</caption>
 * import { useSavedPayments } from '@magento/peregrine/lib/talons/SavedPaymentsPage/useSavedPaymentsPage';
 */
export const useSavedPaymentsPage = (props = {}) => {
    const operations = mergeOperations(defaultOperations, props.operations);
    const { getSavedPaymentsQuery } = operations;

    const [
        ,
        {
            actions: { setPageLoading }
        }
    ] = useAppContext();
    const history = useHistory();
    const [{ isSignedIn }] = useUserContext();

    const { data: savedPaymentsData, loading } = useQuery(
        getSavedPaymentsQuery,
        {
            fetchPolicy: 'cache-and-network',
            nextFetchPolicy: 'cache-first',
            skip: !isSignedIn
        }
    );

    // If the user is no longer signed in, redirect to the home page.
    useEffect(() => {
        if (!isSignedIn) {
            history.push('/');
        }
    }, [history, isSignedIn]);

    // Update the page indicator if the GraphQL query is in flight.
    useEffect(() => {
        setPageLoading(loading);
    }, [loading, setPageLoading]);

    const savedPayments = normalizeTokens(savedPaymentsData);

    return {
        isLoading: loading,
        savedPayments
    };
};

/** JSDoc type definitions */

/**
 * GraphQL formatted string queries used in this talon.
 *
 * @typedef {Object} SavedPaymentsPageQueries
 *
 * @property {GraphQLAST} getSavedPaymentsQuery Query for getting saved payments. See https://devdocs.magento.com/guides/v2.4/graphql/queries/customer-payment-tokens.html
 *
 * @see [savedPaymentsPage.gql.js]{@link https://github.com/magento/pwa-studio/blob/develop/packages/peregrine/lib/talons/SavedPaymentsPage/savedPaymentsPage.gql.js}
 * for queries used in Venia
 */

/**
 * Props data to use when rendering a cart page component.
 *
 * @typedef {Object} SavedPaymentsPageTalonProps
 *
 * @property {function} handleAddPayment Callback function to add a payment.
 * @property {boolean} isLoading true if the query is refreshing from network
 * @property {Array<Object>} savedPayments  An array of saved payment data.
 */
