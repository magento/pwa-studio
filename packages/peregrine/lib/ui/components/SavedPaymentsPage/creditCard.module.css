.root {
    composes: border-2 from global;
    composes: border-solid from global;
    composes: border-subtle from global;
    composes: gap-x-xs from global;
    composes: grid from global;
    composes: grid-cols-autoLast from global;
    composes: min-w-[20rem] from global;
    composes: px-md from global;
    composes: py-sm from global;
    composes: rounded-md from global;
    composes: relative from global;
    min-height: 10rem;
}

.root_active {
    composes: root;

    composes: border-brand-dark from global;
}

.title {
    composes: font-semibold from global;
    grid-column: 1 / span 1;
    grid-row: 1 / span 1;
}

.number {
    grid-column: 1 / span 1;
    grid-row: 2 / span 1;
}

.expiry_date {
    grid-column: 1 / span 1;
    grid-row: 3 / span 1;
}

.delete {
    grid-column: 2 / span 1;
    grid-row: 1 / span 3;
}

.deleteButton {
    composes: root from '../LinkButton/linkButton.module.css';

    composes: no-underline from global;
}

.deleteConfirmationContainer {
    composes: absolute from global;
    composes: bg-white from global;
    composes: gap-md from global;
    composes: grid from global;
    composes: h-full from global;
    composes: items-center from global;
    composes: justify-items-center from global;
    composes: left-0 from global;
    composes: opacity-100 from global;
    composes: px-md from global;
    composes: py-xs from global;
    composes: rounded-md from global;
    composes: top-0 from global;
    composes: w-full from global;
    grid-template-columns: repeat(auto-fit, minmax(10rem, 1fr));
    transform: scaleX(1);
    transition-duration: 192ms;
    transition-timing-function: var(--venia-global-anim-in);
    transition-property: opacity, transform, visibility;

    /* TODO @TW: review (B6) */
    /* composes: visible from global; */
    visibility: visible;
}

.deleteConfirmationContainer_hidden {
    composes: deleteConfirmationContainer;

    composes: opacity-0 from global;
    transform: scaleX(0);
    transition-delay: 0s, 192ms, 0s;
    transition-timing-function: var(--venia-global-anim-out);

    /* TODO @TW: review (B6) */
    /* composes: invisible from global; */
    visibility: hidden;
}

.confirmDeleteButton {
    composes: root_normalPriorityNegative from '../Button/button.module.css';

    composes: bg-white from global;
    composes: order-1 from global;
    composes: order-first from global;
}

.cancelDeleteButton {
    composes: root_lowPriority from '../Button/button.module.css';

    composes: bg-white from global;
}

.deleteText {
    composes: hidden from global;

    composes: xs_lg_block from global;
}

@media (max-width: 479px) {
    .deleteConfirmationContainer {
        transform: scaleY(1);
    }

    .deleteConfirmationContainer_hidden {
        transform: scaleY(0);
    }
}
