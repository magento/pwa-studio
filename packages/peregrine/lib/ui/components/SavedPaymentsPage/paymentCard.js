import React from 'react';
import { shape, string } from 'prop-types';

import paymentCardMapper from './savedPaymentTypes';

const PaymentCard = props => {
    const PaymentComponent = paymentCardMapper[props.payment_method_code];

    if (!PaymentComponent) {
        /**
         * Will be handled in https://jira.corp.magento.com/browse/PWA-1202
         */
    }

    return <PaymentComponent {...props} />;
};

export default PaymentCard;

PaymentCard.propTypes = {
    details: shape({
        expirationDate: string,
        maskedCC: string,
        type: string
    }),
    payment_method_code: string.isRequired
};
