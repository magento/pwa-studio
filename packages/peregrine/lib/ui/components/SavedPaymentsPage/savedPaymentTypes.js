/**
 * This file is augmented at build time using the @magento/venia-ui build
 * target "savedPaymentTypes", which allows third-party modules to add new saved payment component mappings.
 *
 * @see [SavedPayment definition object]{@link SavedPaymentDefinition}
 */
export default {};

/**
 * A payment definition object that describes a saved payment in your storefront.
 *
 * @typedef {Object} SavedPaymentDefinition
 * @property {string} paymentCode is use to map your payment
 * @property {string} importPath Resolvable path to the component the
 *   Route component will render
 *
 * @example <caption>A custom payment method</caption>
 * const myCustomSavedPayment = {
 *      paymentCode: 'cc',
 *      importPath: '@partner/module/path_to_your_component'
 * }
 */
