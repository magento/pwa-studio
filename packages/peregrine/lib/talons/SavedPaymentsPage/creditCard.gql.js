import { gql } from '@apollo/client';

import { SavedPaymentsFragment } from './savedPaymentsPage.gql';

export const DELETE_CREDIT_CARD_PAYMENT = gql`
    mutation DeleteCreditCardPayment($paymentHash: String!) {
        deletePaymentToken(public_hash: $paymentHash) {
            customerPaymentTokens {
                ...SavedPaymentsFragment
            }
            result
        }
    }
    ${SavedPaymentsFragment}
`;

export default {
    deleteCreditCardPaymentMutation: DELETE_CREDIT_CARD_PAYMENT
};
