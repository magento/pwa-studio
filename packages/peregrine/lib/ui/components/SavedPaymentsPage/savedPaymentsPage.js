import React, { useMemo } from 'react';
import { useIntl } from 'react-intl';

import { useSavedPaymentsPage } from '@magento/peregrine/lib/talons/SavedPaymentsPage/useSavedPaymentsPage';
import { useStyle } from '@magento/venia-ui/lib/classify';

import { StoreTitle } from '../Head';
import PaymentCard from './paymentCard';
import { fullPageLoadingIndicator } from '../LoadingIndicator';

import defaultClasses from './savedPaymentsPage.module.css';

const SavedPaymentsPage = props => {
    const talonProps = useSavedPaymentsPage();

    const { isLoading, savedPayments } = talonProps;

    const classes = useStyle(defaultClasses, props.classes);

    const { formatMessage } = useIntl();

    const savedPaymentElements = useMemo(() => {
        if (savedPayments.length) {
            return savedPayments.map(paymentDetails => (
                <PaymentCard
                    key={paymentDetails.public_hash}
                    {...paymentDetails}
                />
            ));
        } else {
            return null;
        }
    }, [savedPayments]);

    const noSavedPayments = useMemo(() => {
        if (!savedPayments.length) {
            return formatMessage({
                id: 'savedPaymentsPage.noSavedPayments',
                defaultMessage: 'You have no saved payments.'
            });
        } else {
            return null;
        }
    }, [savedPayments, formatMessage]);

    const title = formatMessage({
        id: 'savedPaymentsPage.title',
        defaultMessage: 'Saved Payments'
    });

    if (isLoading) {
        return fullPageLoadingIndicator;
    }

    return (
        <div className={classes.root}>
            <StoreTitle>{title}</StoreTitle>
            <h1 className={classes.heading}>{title}</h1>
            <div className={classes.content}>{savedPaymentElements}</div>
            <div className={classes.noPayments}>{noSavedPayments}</div>
        </div>
    );
};

export default SavedPaymentsPage;
