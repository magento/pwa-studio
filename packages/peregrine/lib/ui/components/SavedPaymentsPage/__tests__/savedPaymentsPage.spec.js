import React from 'react';
import { createTestInstance } from '@magento/peregrine';
import { useSavedPaymentsPage } from '@magento/peregrine/lib/talons/SavedPaymentsPage/useSavedPaymentsPage';

import SavedPaymentsPage from '../savedPaymentsPage';

jest.mock('@magento/venia-ui/lib/classify');

jest.mock('../../Head', () => ({ StoreTitle: () => 'Title' }));
jest.mock(
    '@magento/peregrine/lib/talons/SavedPaymentsPage/useSavedPaymentsPage',
    () => {
        return {
            useSavedPaymentsPage: jest.fn()
        };
    }
);
jest.mock('../paymentCard', () => props => <mock-PaymentCard {...props} />);

const props = {};
const talonProps = {
    savedPayments: [
        {
            public_hash: '78asfg87ibafv',
            payment_method_code: 'braintree',
            details: {
                maskedCC: '1234',
                type: 'VI',
                expirationDate: '12/12/2022'
            }
        }
    ]
};

test('renders correctly when there are no existing saved payments', () => {
    // Arrange.
    useSavedPaymentsPage.mockReturnValueOnce({ savedPayments: [] });

    // Act.
    const instance = createTestInstance(<SavedPaymentsPage {...props} />);

    // Assert.
    expect(instance.toJSON()).toMatchSnapshot();
});

test('renders correctly when there are existing saved payments', () => {
    // Arrange.
    useSavedPaymentsPage.mockReturnValueOnce(talonProps);

    // Act.
    const instance = createTestInstance(<SavedPaymentsPage {...props} />);

    // Assert.
    expect(instance.toJSON()).toMatchSnapshot();
});

test('renders loading indicator when isLoading is true', () => {
    // Arrange.
    useSavedPaymentsPage.mockReturnValueOnce({
        ...talonProps,
        isLoading: true
    });

    // Act.
    const instance = createTestInstance(<SavedPaymentsPage {...props} />);

    // Assert.
    expect(instance.toJSON()).toMatchSnapshot();
});
