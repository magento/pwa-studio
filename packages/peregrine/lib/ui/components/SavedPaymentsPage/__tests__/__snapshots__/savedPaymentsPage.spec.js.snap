// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when there are existing saved payments 1`] = `
<div
  className="root"
>
  Title
  <h1
    className="heading"
  >
    Saved Payments
  </h1>
  <div
    className="content"
  >
    <mock-PaymentCard
      details={
        Object {
          "expirationDate": "12/12/2022",
          "maskedCC": "1234",
          "type": "VI",
        }
      }
      payment_method_code="braintree"
      public_hash="78asfg87ibafv"
    />
  </div>
  <div
    className="noPayments"
  />
</div>
`;

exports[`renders correctly when there are no existing saved payments 1`] = `
<div
  className="root"
>
  Title
  <h1
    className="heading"
  >
    Saved Payments
  </h1>
  <div
    className="content"
  />
  <div
    className="noPayments"
  >
    You have no saved payments.
  </div>
</div>
`;

exports[`renders loading indicator when isLoading is true 1`] = `
<div
  className="global"
>
  <span
    className="root"
  >
    <svg
      className="icon"
      fill="none"
      height={64}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width={64}
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="12"
        x2="12"
        y1="2"
        y2="6"
      />
      <line
        x1="12"
        x2="12"
        y1="18"
        y2="22"
      />
      <line
        x1="4.93"
        x2="7.76"
        y1="4.93"
        y2="7.76"
      />
      <line
        x1="16.24"
        x2="19.07"
        y1="16.24"
        y2="19.07"
      />
      <line
        x1="2"
        x2="6"
        y1="12"
        y2="12"
      />
      <line
        x1="18"
        x2="22"
        y1="12"
        y2="12"
      />
      <line
        x1="4.93"
        x2="7.76"
        y1="19.07"
        y2="16.24"
      />
      <line
        x1="16.24"
        x2="19.07"
        y1="7.76"
        y2="4.93"
      />
    </svg>
  </span>
  <span
    className="message"
  >
    <mock-FormattedMessage
      defaultMessage="Fetching Data..."
      id="loadingIndicator.message"
    />
  </span>
</div>
`;
